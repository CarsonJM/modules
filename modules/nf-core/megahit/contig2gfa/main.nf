process MEGAHIT_CONTIG2GFA {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/megahit_pigz:657d77006ae5f222' :
        'community.wave.seqera.io/library/megahit_pigz:87a590163e594224' }"

    input:
    tuple val(meta), path(megahit_contigs)

    output:
    tuple val(meta), path("${prefix}.graph.fastg.gz")   , emit: fastg
    tuple val(meta), path("${prefix}.gfa.gz")           , emit: gfa
    path "versions.yml"                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args            = task.ext.args ?: ''
    def args2           = task.ext.args2 ?: ''
    def is_compressed   = megahit_contigs.getExtension() == "gz" ? true : false
    def contigs_name    = is_compressed ? megahit_contigs.getBaseName() : megahit_contigs
    prefix              = task.ext.prefix ?: "${meta.id}"
    """
    # decompress megahit_contigs if compressed
    if [ "${is_compressed}" == "true" ]; then
        gzip -c -d ${megahit_contigs} > ${contigs_name}
    fi

    # identify kmer size used to create assemblies
    kmer_size=\$(grep "^>" ${contigs_name} | sed 's/>k//; s/_.*//' | head -n 1)

    # convert contigs to FastG format
    megahit_toolkit \\
        contig2fastg \\
        \$kmer_size \\
        ${contigs_name} > ${prefix}.graph.fastg

    # convert FastG to GFA format
    fastg2gfa \\
        ${prefix}.graph.fastg > ${prefix}.gfa

    gzip ${prefix}.gfa ${prefix}.graph.fastg

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        megahit: \$(echo \$(megahit -v 2>&1) | sed 's/MEGAHIT v//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def is_compressed   = megahit_contigs.getExtension() == "gz" ? true : false
    def contigs_name    = is_compressed ? megahit_contigs.getBaseName() : megahit_contigs
    prefix              = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}.graph.fastg.gz
    echo "" | gzip > ${prefix}.gfa.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        megahit: \$(echo \$(megahit -v 2>&1) | sed 's/MEGAHIT v//')
    END_VERSIONS
    """
}
