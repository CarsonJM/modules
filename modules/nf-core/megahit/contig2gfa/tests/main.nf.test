nextflow_process {

    name "Test Process MEGAHIT_CONTIG2GFA"
    script "../main.nf"
    process "MEGAHIT_CONTIG2GFA"

    tag "modules"
    tag "modules_nfcore"
    tag "megahit"
    tag "megahit/contig2gfa"

    test("sarcov2 - fasta") {
        when {
            process {
                """
                input[0] = [
                    [id:"test", single_end:true],
                    file("${projectDir}/assets/test-datasets/genomics/sarscov2/illumina/fasta/test.megahit.contigs.fa.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [id:"test", single_end:true],
                    file("${projectDir}/assets/test-datasets/genomics/sarscov2/illumina/fasta/test.megahit.contigs.fa.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
