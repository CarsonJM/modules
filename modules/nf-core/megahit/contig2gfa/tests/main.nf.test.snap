{
    "sarcov2 - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.graph.fastg.gz:md5,c0d204a1f25ad863d2a16b6327625038"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.gfa.gz:md5,617ff12a7a83eebe21ce38eb199309d5"
                    ]
                ],
                "2": [
                    "versions.yml:md5,a2b4a72a5fcf97309f295e2a8585c356"
                ],
                "fastg": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.graph.fastg.gz:md5,c0d204a1f25ad863d2a16b6327625038"
                    ]
                ],
                "gfa": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.gfa.gz:md5,617ff12a7a83eebe21ce38eb199309d5"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a2b4a72a5fcf97309f295e2a8585c356"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T13:42:27.376982934"
    },
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.graph.fastg.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.gfa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "2": [
                    "versions.yml:md5,a2b4a72a5fcf97309f295e2a8585c356"
                ],
                "fastg": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.graph.fastg.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "gfa": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.gfa.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a2b4a72a5fcf97309f295e2a8585c356"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T13:42:36.239344606"
    }
}